src/linux/cpu.cpp

```diff
@@ -10,29 +10,173 @@
 #include <map>
 #include <string>
 #include <vector>
+#include <unistd.h>
+#include <cmath>
+
+#include <iostream>
+#include <sstream>
+#include <cstdlib>
+#include <cstdio>
 
 #include "hwinfo/cpu.h"
 #include "hwinfo/utils/stringutils.h"
+#include "hwinfo/utils/filesystem.h"
 
 namespace hwinfo {
 
-int64_t get_freq_by_id_and_type(int proc_id, const std::string& type) {
-  std::string line;
-  std::ifstream stream("/sys/devices/system/cpu/cpu" + std::to_string(proc_id) + "/cpufreq/" + type);
-  if (!stream) {
-    return -1;
+// _____________________________________________________________________________________________________________________
+int64_t getMaxClockSpeed_MHz(const int& core_id) {
+  int64_t Hz = filesystem::get_specs_by_file_path("/sys/devices/system/cpu/cpu" + std::to_string(core_id) + "/cpufreq/scaling_max_freq");
+  if (Hz > -1)
+  {
+    return Hz / 1000;
   }
-  getline(stream, line);
-  stream.close();
-  try {
-    return std::stoll(line) / 1000;
-  } catch (std::invalid_argument& e) {
-    return -1;
+
+  return -1;
+}
+
+// _____________________________________________________________________________________________________________________
+int64_t getRegularClockSpeed_MHz(const int& core_id) {
+  int64_t Hz = filesystem::get_specs_by_file_path("/sys/devices/system/cpu/cpu" + std::to_string(core_id) + "/cpufreq/base_frequency");
+  if (Hz > -1)
+  {
+    return Hz / 1000;
   }
+
+  return -1;
+}
+
+int64_t getMinClockSpeed_MHz(const int& core_id) {
+  int64_t Hz = filesystem::get_specs_by_file_path("/sys/devices/system/cpu/cpu" + std::to_string(core_id) + "/cpufreq/scaling_min_freq"); 
+  if (Hz > -1)
+  {
+    return Hz / 1000;
+  }
+
+  return -1;
 }
 
 // _____________________________________________________________________________________________________________________
-int64_t CPU::currentClockSpeed_MHz() const { return get_freq_by_id_and_type(_core_id, "scaling_cur_freq"); }
+int64_t CPU::currentClockSpeed_MHz() const {
+  int64_t Hz = filesystem::get_specs_by_file_path("/sys/devices/system/cpu/cpu" + std::to_string(_core_id) + "/cpufreq/scaling_cur_freq");
+  if (Hz > -1)
+  {
+    return Hz / 1000;
+  }
+
+  return -1;
+ }
+
+double CPU::currentUtility_Percentage() const {
+  if (!_initialize_jiffies)
+  {
+    // Sleep 1 sec just for the start cause the usage needs to have a delta value which is depending on the unix file read
+    // it's just for the init, you don't need to wait if the delta is already created ...
+    usleep(1000 * 1000);
```

Comment:
Must it be a full second? I have no experience with this...

Reply:
It's basically the write-time linux needs to update the stat file .... it may work with 50ms too if the user is having a good SSD .... again it's just for the 1st time the function getting called, what I can do is doing a ranged while loop which checks the file every 50ms and detects if the file changed, if so run the calculation, otherwise just let the ranged loop end and return the usual value which is most likely the previous value or -1

Reply:
Okay... I'm just considering how to make this clearer to users. I'm imagine the scenario where a user calls the function the first time in a runtime critical context.

Maybe it is best to add a init_jiffies() that does the initialisation and can be called somewhere else. What do you think?

Reply:
I can do this, however, it's just cosmetic, `get_jiffies` will work anyway with or without initialization, the thing is, it will return just -1 until linux decides to update the file, but the file surely will update regardless of the state of HWInfo ... there is nothing wrong with returning -1 for some amount of milliseconds after `getAllSockets` got called ... might be better to add a `bool one_frame_init` in `getAllSockets` which does await the stats file, and maybe does other future initialization stuff, I'm fine with both ways

Reply:
Ah okay! You are correct, there is no point then. Maybe a Boolean argument stating to wait or not to wait?